docs/async-use-cases.md

```diff
@@ -0,0 +1,95 @@
+    `Jexl.eval()`.
+  - Network requests where the HTTP request happens on the foreign side, then gets parsed on the
+    Rust side, then the parsed result is returned to the foreign side.
+
+## Normal call / Foreign thread / Thread switch / Blocking
+
+This means the foreign code wants to execute something on a Rust thread.  This one seems unlikely to
+me, but here's one possible use case: Rust starts up a tokio event loop, then
+the foreign code blocks on an HTTP request that runs in that event loop.
+
+## Normal call / Foreign thread / Thread switch / Async
+
+This is basically the same as the previous section, except the foreign side executes the call async
+rather than blocking on it which seems more likely in the real world.
+
+## Normal call / Rust Thread / * / *
+
+In general, it's not safe for the foreign code to run on a Rust-controlled thread.  Some languages
+may support it, but it's definitely not going to work on Desktop JS.  I think we should consider
+this invalid.
+
+## CallbackInterface / Foreign thread / Same thread / Blocking
+
+This is a normal CallbackInterace call and is already supported.
+
+## CallbackInterface / Foreign thread / Same thread / Async
+
+Async call from Rust to the foreign language.  Some use cases:
+   - Nimbus JEXL evaluation
+   - Async viaduct call
+
+Since this is happening on a foreign thead, it means there's a foreign caller that's waiting for a
+return.  There's a couple ways to handle that:
+
+  - Rust returns void, which is the current plan for Nimbus.  This doesn't mean it's a no-op, since
+    when the async calls resolve the Rust code will continue to do work.
+  - The Rust call is async, so it returns a Future/Promise (this is the other side of the
+    `async Nimbus.init()` use case described above).
+
+## CallbackInterface / Foreign thread / Thread switch / * and CallbackInterface / Rust Thread / Same thread / *
+
+This is invalid because it's not safe for foreign code to run on a Rust thread (see above).
+
+## CallbackInterface / Rust Thread / Thread switch / Blocking
+
+A Rust thread wants to make a blocking call into the foreign code, which requires a thread switch.
+For example, Rust creates a thread pool of workers.  Sometimes those workers want to make network
+calls, so they make a viaduct call.  Since this is happening in a thread pool, it's fine for the
+thread to block on the result.
+
+## CallbackInterface / Rust Thread / Thread switch / Async
+
+This is the same as the last one, except the call is async.  The main use case is probably the same
+as the previous section, except when Rust is using an event loop rather than a thread pool.
+
+# New features
+
+Based on the above analysis, I believe there are 2 features that we should consider adding:
+
+  - Rust code making an async call into a CallbackInterface / Foreign code
+    making an async into Rust. I think this could be implemented by having
+    UniFFI generate something like the hand-written demo from #1252.
```

Comment:
This seems like a nice feature, but I'm not sure it's necessary.  I think we could implement the Nimbus work with hand-written code.  Maybe we could just start there and see if there's more demand before generating that code with UniFFI.

Reply:
I agree that hand-writing & using it that way seems sensible. That's how UniFFI was born to begin with.